### PR TITLE
DBZ-1367 DBZ-1368 DBZ-1399 Toasted column not read out-of-bands

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-1367
https://issues.jboss.org/browse/DBZ-1368
https://issues.jboss.org/browse/DBZ-1399

Depends on https://github.com/debezium/postgres-decoderbufs/pull/20
The code now expects that bot pgoutput and protobuf signals fields that contain unchanged toasted values. wal2json with `columns_diff_exclude_unchanged_toast` supposes that missing toasted column means exactly the same thing.
The code replaces missing value with a string or binary placeholder. In case of update messages when non-toasted column is updated the PR will read old toasted value if available in the message and write it into the new value instead of placeholder.

### CODE REVIEW

**Comment by Naros** [comment]
PR #1028 can be merged after this PR has been integrated.  See that PR for details.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java [hunk 1]
```diff
@@ -44,6 +48,9 @@
     private final PostgresConnectorConfig connectorConfig;
     private final PostgresConnection connection;
     private final TableId tableId;
+    private final boolean isJsonPlugin;
+    private Object[] cachedOldColumnValues;
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: @jpechane What's the purpose of this "cache"? This warrants a field-level comment for sure.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java [hunk 2]
```diff
@@ -54,6 +61,7 @@ public PostgresChangeRecordEmitter(OffsetContext offset, Clock clock, PostgresCo
         this.connection = connection;
 
         this.tableId = PostgresSchema.parse(message.getTable());
+        this.isJsonPlugin = "wal2json".equals(connectorConfig.plugin().getPostgresPluginName());
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: Instead of checking the plug-in name, I'd rather suggest adding a method to `LogicalDecoder` that describes what you're actually after, e.g. `conveysUnchangedToastColumnValueMarker()` which would return true for pgoutput and decoderbufs and false for wal2json.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java [hunk 3]
```diff
@@ -18,6 +18,10 @@
  */
 public class ToastedReplicationMessageColumn extends AbstractReplicationMessageColumn {
 
+    public static enum ToastedValue {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java]: What's the purpose of this enum with a single value? It's a bit unclear.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java [hunk 4]
```diff
@@ -89,11 +89,6 @@ public ChainedLogicalStreamBuilder optionsWithoutMetadata(ChainedLogicalStreamBu
             .withSlotOption("include-timestamp", 1);
     }
 
-    @Override
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java]: Why is this removed? Doesn't it mean we'd get the unchanged TOAST column values again for some wal2json versions on RDS?

File: .travis.yml [hunk 5]
```diff
@@ -5,9 +5,9 @@ cache:
 
 env:
   - MAVEN_CLI: '"clean install -B -pl debezium-connector-mysql -am -Passembly"'
-  - MAVEN_CLI: '"clean install -B -pl debezium-connector-postgres -am -Passembly"'
-  - MAVEN_CLI: '"clean install -B -pl debezium-connector-postgres -am -Passembly,wal2json-decoder"'
-  - MAVEN_CLI: '"clean install -B -pl debezium-connector-postgres -am -Passembly,pgoutput-decoder,postgres-10 -Ddebezium.test.records.waittime=5"'
+  - MAVEN_CLI: '"clean install -B -pl debezium-connector-postgres -am -Passembly -Dversion.postgres.server=10-dev"'
```
[the remark this answers is not in the document]
Comment (jpechane) [reply .travis.yml]: @gunnarmorling That's image build from associated protobuf PR. It is necessary to do it this way to make sure that the tests are passing on Travis

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java (hunk 2 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: @gunnarmorling of course, sorry for that

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java (hunk 4 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java]: @gunnarmorling We are not able to make this running correctly - see https://issues.jboss.org/browse/DBZ-1399

**Comment by gunnarmorling** [comment]
@jpechane, there should also be a documentation update part of this PR; apart from the new connector option, the handling of unchanged TOASTed columns should also be described for each of the supported plug-ins.

**Comment by gunnarmorling** [comment]
> the PR will read old toasted value if available in the message and write it into the new value instead of placeholder.

Where is this happening? Can you point me to it? Btw. didn't you mean to use an SMT for that originally?

**Comment by jpechane** [comment]
@gunnarmorling See https://github.com/debezium/debezium/pull/1031/files#diff-11eb08684131432b99e3565263fbfa5bR175
It is not read from the database at all.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java (hunk 3 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java]: @gunnarmorling This one is used in value converters. When detected it is replaced with the placeholder value depending on target datatype - either string or binary

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java (hunk 2 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: @gunnarmorling Changed

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java (hunk 1 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: @gunnarmorling  Comment added

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply .travis.yml]: Ok, so should this be reverted then?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 6]
```diff
@@ -729,6 +754,16 @@ public static SchemaRefreshMode parse(String value) {
...
             .withValidation(Field::isNonNegativeLong);
 
+    public static final Field TOASTED_VALUE_PLACEHOLDER = Field.create("toasted.value.placeholder")
+            .withDisplayName("Toasted value placeholder")
+            .withType(Type.STRING)
+            .withWidth(Width.MEDIUM)
+            .withDefault("__DEBEZIUM_TOASTED_VALUE__")
+            .withImportance(Importance.MEDIUM)
+            .withDescription("Specify the constant that will be provided by Debezium to indicate that " +
+                    "the original value is a toasted value not provided by the database." +
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java]:
```suggestion
                    "the original value is an unchanged toasted value not provided by the database." + 
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java]: Ok, got it. Why is it not a simple constant then:

    public static final Object UNCHANGED_TOAST_MARKER = new Object();

And in the code above you could ask `if data == UNCHANGED_TOAST_MARKER ...`. Using an enum implies that there are multiple values really.

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply .travis.yml]: Tracked as https://issues.jboss.org/browse/DBZ-1504

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java (hunk 4 above)
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java]: Let's add the implication of this to the notes on RDS.

Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/wal2json/NonStreamingWal2JsonMessageDecoder.java]: @gunnarmorling Docs added

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java (hunk 1 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresChangeRecordEmitter.java]: @gunnarmorling Removed

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java (hunk 3 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/ToastedReplicationMessageColumn.java]: @gunnarmorling Done

File: documentation/modules/ROOT/pages/connectors/postgresql.adoc [hunk 7]
```diff
@@ -1659,6 +1685,12 @@ See xref:configuration/avro.adoc#names[Avro naming] for more details.
 |10000 (10 seconds)
 |The number of milli-seconds to wait between retry attempts when the connector fails to connect to a replication slot.
 
+|`toasted.value.placeholder`
+|`__DEBEZIUM_TOASTED_VALUE__`
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/connectors/postgresql.adoc]: @jpechane, one last comment: How about changing the default value into something more neutral, e.g. "__debezium_unavailable_value"? That'd allow for potential re-use later on, e.g. in the context of MySQL's non-full binlog mode. Also for consumers the fact that this is from a TOAST column is mostly a technicality. Note I'd also suggest to lower-case it and only have "__" at the beginning to be consistent with other usages of "__debezium..." literals.
